File: fedadam.py
import numpy as np
from model import make_model


def fedadam(x_train, y_train, x_test, y_test, clients=3, rounds=3, local_epochs=1, 
            batch_size=64, server_lr=0.01, beta1=0.9, beta2=0.999, tau=1e-3, 
            full_data_per_client=False):
    """FedAdam: Adaptive Federated Optimization with server-side Adam.
    
    Uses momentum (beta1) and adaptive learning rates (beta2) on the server
    to update the global model based on the pseudo-gradient (difference between
    current and averaged client models).
    
    Args:
        server_lr: Server learning rate (eta in paper)
        beta1: Exponential decay rate for first moment estimates
        beta2: Exponential decay rate for second moment estimates
        tau: Small constant for numerical stability
    
    Returns:
        Test accuracy (float)
    """
    n = len(x_train)
    if full_data_per_client:
        shards = [np.arange(n) for _ in range(clients)]
    else:
        idx = np.random.permutation(n)
        shards = np.array_split(idx, clients)
    
    # Initialize global model
    global_model = make_model()
    global_weights = global_model.get_weights()
    
    # Initialize Adam moments (server-side)
    m_t = [np.zeros_like(w) for w in global_weights]  # First moment
    v_t = [np.zeros_like(w) for w in global_weights]  # Second moment
    
    for t in range(1, rounds + 1):
        client_weights = []
        
        for shard in shards:
            if len(shard) == 0:
                continue
            
            local = make_model()
            local.set_weights(global_weights)
            local.fit(x_train[shard], y_train[shard], epochs=local_epochs, 
                     batch_size=batch_size, verbose=0)
            client_weights.append(local.get_weights())
        
        if len(client_weights) == 0:
            break
        
        # Average client weights
        avg_weights = []
        for weights in zip(*client_weights):
            avg_weights.append(np.mean(weights, axis=0))
        
        # Compute pseudo-gradient: delta = w_t - avg_weights
        delta = [w_old - w_new for w_old, w_new in zip(global_weights, avg_weights)]
        
        # Update moments and global weights using Adam
        new_global_weights = []
        for i, (w, d, m, v) in enumerate(zip(global_weights, delta, m_t, v_t)):
            # Update biased first moment estimate
            m_new = beta1 * m + (1 - beta1) * d
            # Update biased second moment estimate
            v_new = beta2 * v + (1 - beta2) * (d ** 2)
            
            # Bias correction
            m_hat = m_new / (1 - beta1 ** t)
            v_hat = v_new / (1 - beta2 ** t)
            
            # Update weights
            w_new = w - server_lr * m_hat / (np.sqrt(v_hat) + tau)
            
            new_global_weights.append(w_new)
            m_t[i] = m_new
            v_t[i] = v_new
        
        global_weights = new_global_weights
    
    # Final evaluation
    global_model.set_weights(global_weights)
    loss, acc = global_model.evaluate(x_test, y_test, verbose=0)
    return float(acc)

File: src/matlab/fedadam.m
function acc = fedadam(x_train, y_train, x_test, y_test, varargin)
% FedAdam: Server-side Adam optimizer for global aggregation.
% Optional name-value:
%  'clients' (3), 'rounds' (3), 'local_epochs' (1), 'batch_size' (64),
%  'server_lr' (0.01), 'beta1' (0.9), 'beta2' (0.999), 'tau' (1e-3),
%  'full_data_per_client' (false), 'report' (function)

    p = inputParser;
    addParameter(p, 'clients', 3);
    addParameter(p, 'rounds', 3);
    addParameter(p, 'local_epochs', 1);
    addParameter(p, 'batch_size', 64);
    addParameter(p, 'server_lr', 0.01);
    addParameter(p, 'beta1', 0.9);
    addParameter(p, 'beta2', 0.999);
    addParameter(p, 'tau', 1e-3);
    addParameter(p, 'full_data_per_client', false);
    addParameter(p, 'report', []);
    parse(p, varargin{:});

    clients = p.Results.clients;
    rounds = p.Results.rounds;
    local_epochs = p.Results.local_epochs;
    batch_size = p.Results.batch_size;
    server_lr = p.Results.server_lr;
    beta1 = p.Results.beta1;
    beta2 = p.Results.beta2;
    tau = p.Results.tau;
    full_data_per_client = p.Results.full_data_per_client;
    report = p.Results.report;

    n = size(x_train, 1);
    shards = make_shards(n, clients, full_data_per_client);

    global_model = make_model();
    global_weights = global_model.get_weights();
    comm_cost = 0;
    prev_acc = [];
    prev_global_weights = copy_weights(global_weights);

    % Server moments
    m_t = cell(size(global_weights));
    v_t = cell(size(global_weights));
    for j = 1:numel(global_weights)
        m_t{j} = zeros(size(global_weights{j}));
        v_t{j} = zeros(size(global_weights{j}));
    end

    for t = 1:rounds
        start_time = tic;
        client_weights = {};
        for s = 1:numel(shards)
            shard = shards{s}; if isempty(shard), continue; end
            local = make_model();
            local.set_weights(global_weights);
            local.fit(x_train(shard,:,:), y_train(shard), 'epochs', local_epochs, 'batchSize', batch_size);
            client_weights{end+1} = local.get_weights(); %#ok<AGROW>
        end
        if isempty(client_weights), break; end
        avg_weights = average_weights(client_weights);
        % Pseudo-gradient
        delta = cell(size(global_weights));
        for j = 1:numel(global_weights)
            delta{j} = global_weights{j} - avg_weights{j};
        end
        % Adam update on server
        new_global_weights = cell(size(global_weights));
        for j = 1:numel(global_weights)
            m_new = beta1 * m_t{j} + (1 - beta1) * delta{j};
            v_new = beta2 * v_t{j} + (1 - beta2) * (delta{j}.^2);
            m_hat = m_new ./ (1 - beta1^t);
            v_hat = v_new ./ (1 - beta2^t);
            new_global_weights{j} = global_weights{j} - server_lr * m_hat ./ (sqrt(v_hat) + tau);
            m_t{j} = m_new; v_t{j} = v_new;
        end
        global_weights = new_global_weights;
        global_model.set_weights(global_weights);

        [metrics, prev_acc, prev_global_weights, comm_cost] = compute_round_metrics('FedAdam', t, start_time, global_model, x_test, y_test, prev_acc, prev_global_weights, global_weights, clients, comm_cost, 'local_weights_list', client_weights, 'reference_weights', avg_weights);
        if ~isempty(report), report('FedAdam', t, metrics); end
    end

    global_model.set_weights(global_weights);
    [~, acc] = global_model.evaluate(x_test, y_test);
end

function new_weights = average_weights(client_weights)
    K = numel(client_weights);
    L = numel(client_weights{1});
    new_weights = cell(1, L);
    for j = 1:L
        sumW = 0; for k = 1:K, sumW = sumW + client_weights{k}{j}; end
        new_weights{j} = sumW / K;
    end
end

function out = copy_weights(weights)
    out = cell(size(weights));
    for i = 1:numel(weights), out{i} = weights{i}; end
end
